File: Day_8_Treetop_Tree_House/Python/main.py
import numpy as np
import numpy.typing as npt
from typing import Any


def is_cumulative_max(A: npt.NDArray[Any]) -> npt.NDArray[bool]:
    """
    Compute whether or not each array element is the cumulative maximum.
    """
    l = A.shape[0]

    return A[1 : l - 1] > np.maximum.accumulate(A[: l - 2])


def n_first_consecutive(A: npt.NDArray[Any]) -> int:
    """
    Compute number of first consecutive occurrences of True.
    """
    where = np.where(np.invert(A))[0]

    # if all elements are true, return last index
    if where.shape[0] == 0:
        return A.shape[0] - 1

    return where[0]


if __name__ == '__main__':
    # file with trees data
    file_path = '../trees.txt'
    # open and read file
    with open(file_path) as f:
        file_contents = f.read()

    # create 2d array of tree heights
    trees = np.array(
        [list(line) for line in file_contents.split() if len(line) > 0],
        dtype=np.int8,
    )
    # get tree grid dimensions
    l, w = trees.shape
    # set of coordinates of non-bordering trees visible from at least one side
    non_bordering_visible = set()

    # iterate over rows
    for i in range(1, l - 1):
        # get trees visible from the left
        visible_from_left = is_cumulative_max(trees[i])
        # get trees visible from the right
        visible_from_right = is_cumulative_max(trees[i, ::-1])

        # perform OR of tree visibilities and update coordinates
        for j in np.where(visible_from_left | visible_from_right[::-1])[0] + 1:
            non_bordering_visible.add((i, j))

    # iterate over columns
    for j in range(1, w - 1):
        # get trees visible from the top
        visible_from_top = is_cumulative_max(trees[:, j])
        # get trees visible from the bottom
        visible_from_bottom = is_cumulative_max(trees[::-1, j])

        # perform OR of tree visibilities and update coordinates
        for i in np.where(visible_from_top | visible_from_bottom[::-1])[0] + 1:
            non_bordering_visible.add((i, j))

    # compute number of trees along the outer border
    # these trees are always visible
    n_bordering_trees = 2 * (l + w - 2)
    # number of all visible trees
    n_visible_trees = len(non_bordering_visible) + n_bordering_trees
    print(n_visible_trees)

    max_scenic_score = 0
    # iterate of rows
    for i in range(l):
        # iterate of columns
        for j in range(w):
            # compute scenic score
            scenic_score = (
                (n_first_consecutive(trees[i, j - 1 :: -1] < trees[i, j]) + 1)
                * (n_first_consecutive(trees[i, j + 1 :] < trees[i, j]) + 1)
                * (n_first_consecutive(trees[i - 1 :: -1, j] < trees[i, j]) + 1)
                * (n_first_consecutive(trees[i + 1 :, j] < trees[i, j]) + 1)
            )

            # update maximum scenic score if current is higher
            if scenic_score > max_scenic_score:
                max_scenic_score = scenic_score

    print(max_scenic_score)

File: Day_8_Treetop_Tree_House/Go/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

// determine whether a single tree is visible from at least one side
func IsVisible(trees [][]int, i int, j int) bool {
	// trees grid dimensions
	l := len(trees)
	w := len(trees[0])

	visible := true
	// check for visibility from the left
	for k := 0; k < j; k++ {
		if trees[i][k] >= trees[i][j] {
			visible = false
			break
		}
	}

	if visible {
		return true
	}

	visible = true
	// check for visibility from the right
	for k := w - 1; k > j; k-- {
		if trees[i][k] >= trees[i][j] {
			visible = false
			break
		}
	}

	if visible {
		return true
	}

	visible = true
	// check for visibility from the top
	for k := 0; k < i; k++ {
		if trees[k][j] >= trees[i][j] {
			visible = false
			break
		}
	}

	if visible {
		return true
	}

	visible = true
	// check for visibility from the bottom
	for k := l - 1; k > i; k-- {
		if trees[k][j] >= trees[i][j] {
			visible = false
			break
		}
	}

	return visible
}

// get scenic score of a single tree
func GetScenicScore(trees [][]int, i int, j int) int {
	// trees grid dimensions
	l := len(trees)
	w := len(trees[0])

	scenicScore := 1
	viewingDistance := 0

	// get viewing distance on the right
	for k := j + 1; k < w; k++ {
		viewingDistance++
		if trees[i][k] >= trees[i][j] {
			break
		}
	}

	scenicScore *= viewingDistance
	viewingDistance = 0

	// get viewing distance on the left
	for k := j - 1; k >= 0; k-- {
		viewingDistance++
		if trees[i][k] >= trees[i][j] {
			break
		}
	}

	scenicScore *= viewingDistance
	viewingDistance = 0

	// get viewing distance on the bottom
	for k := i + 1; k < l; k++ {
		viewingDistance++
		if trees[k][j] >= trees[i][j] {
			break
		}
	}

	scenicScore *= viewingDistance
	viewingDistance = 0

	// get viewing distance on the top
	for k := i - 1; k >= 0; k-- {
		viewingDistance++
		if trees[k][j] >= trees[i][j] {
			break
		}
	}

	scenicScore *= viewingDistance

	return scenicScore
}

// get line-by-line file scanner
func GetFileScanner(filePath string) (*bufio.Scanner, error) {
	f, err := os.Open(filePath)
	if err != nil {
		return nil, fmt.Errorf("failed to open file %s, %s", filePath, err.Error())
	}

	scanner := bufio.NewScanner(f)

	return scanner, nil
}

func main() {
	// file with trees data
	filePath := "../trees.txt"

	scanner, err := GetFileScanner(filePath)
	if err != nil {
		fmt.Println("failed to get file scanner", err.Error())
		return
	}

	// ASCII value of '0' for converting runes to digits
	zeroASCII := int('0')
	// 2-dimensional trees grid
	trees := make([][]int, 0)

	// scan file line-by-line and construct trees grid
	lineIdx := 0
	for scanner.Scan() {
		input := scanner.Text()
		trees = append(trees, make([]int, len(input)))

		for charIdx := range input {
			trees[lineIdx][charIdx] = int(input[charIdx]) - zeroASCII
		}
		lineIdx += 1
	}

	// trees grid dimensions
	l := len(trees)
	w := len(trees[0])

	nonBorderingVisibleCount := 0
	maxScenicScore := 0

	// iterate over non-bordering trees
	for i := 1; i < l-1; i++ {
		for j := 1; j < w-1; j++ {
			// check if tree is visible and increment count
			if IsVisible(trees, i, j) {
				nonBorderingVisibleCount++
			}

			// get scenic score of tree and update max if appropriate
			scenicScore := GetScenicScore(trees, i, j)
			if maxScenicScore < scenicScore {
				maxScenicScore = scenicScore
			}
		}
	}

	// number of trees along the outer border
	// these trees are always visible
	borderingTreesCount := 2 * (l + w - 2)

	fmt.Println(nonBorderingVisibleCount + borderingTreesCount)
	fmt.Println(maxScenicScore)
}
